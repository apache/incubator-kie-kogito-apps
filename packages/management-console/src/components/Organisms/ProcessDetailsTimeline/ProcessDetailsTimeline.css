.processdetailstimetine-iconstyle {
    position: relative;
    top: 3px;
    left: 3px;
}