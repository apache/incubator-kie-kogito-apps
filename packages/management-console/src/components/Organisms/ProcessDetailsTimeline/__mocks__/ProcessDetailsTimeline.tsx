import React from 'react'

const MockedProcessTimeline = (): React.ReactElement => {
    return <></>
}

export default MockedProcessTimeline