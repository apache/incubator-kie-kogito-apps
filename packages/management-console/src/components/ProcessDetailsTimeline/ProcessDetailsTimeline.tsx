import React from 'react';
import { Card, CardHeader, CardBody, CardFooter, Button } from '@patternfly/react-core';
import { TimeAgo } from '@n1ru4l/react-time-ago';
import { UserIcon } from '@patternfly/react-icons'
import { ServicesIcon } from '@patternfly/react-icons'


const ProcessDetailsTimeline = ({ loading, data }) => {
  const TimelineStyle = {
    marginLeft: '2em',
    height: '40em',
    position: 'relative',
    bottom: '3em',
    overflowY: 'scroll',
    marginTop: '3em'
  };
  
  const IconStyle = {
    position: 'relative',
    top: '3px',
    left: '3px'
  }
  return (
    <Card style={TimelineStyle}>
      <CardHeader>Timeline</CardHeader>
      <CardBody>
        <div className="timeline-container">
          {!loading ? (
            data[0].nodes.map((content, index) => {
              return (
                <div className="timeline-item" key={index}>
                  <div className="timeline-item-content">
                      <TimeAgo date={new Date(`${content.exit}`)} render={({ error, value }) => <span>{value}</span>}/>
                    <p>{content.name}</p>
                    <span className="circle">{content.type === 'HumanTaskNode' ? <UserIcon style={IconStyle}/> : <ServicesIcon style={IconStyle}/> }  </span>
                  </div>
                </div>
              );
            })
          ) : (
            <p>loading...</p>
          )}
        </div>
      </CardBody>
      <CardFooter>
        <Button variant="primary" style={{ float: 'right' }}>
          Primary
        </Button>
      </CardFooter>
    </Card>
  );
};

export default ProcessDetailsTimeline;
