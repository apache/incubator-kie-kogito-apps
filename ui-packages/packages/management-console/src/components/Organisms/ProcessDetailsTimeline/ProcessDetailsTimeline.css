/* This adds the line extending the length of the timeline, while the items are pushed left to overlap it halfway. The first child is pushed up to hide the top of the dashed line. The status icon must be given a background so that the line doesn't show through the icon. */
.kogito-management-console--timeline {
    border-left: var(--pf-global--BorderWidth--md) dashed var(--pf-global--BorderColor--300);
    margin-left: var(--pf-global--spacer--sm);
    margin-top: var(--pf-global--spacer--sm);
  }
.kogito-management-console--timeline-item {
  background-color: transparent;
  margin-left: calc(-1 * var(--pf-global--spacer--sm));
}
.kogito-management-console--timeline-item:first-child {
  margin-top: calc(-1 * var(--pf-global--spacer--sm));
}
.kogito-management-console--timeline-status {
  background-color: var(--pf-global--BackgroundColor--100);
}