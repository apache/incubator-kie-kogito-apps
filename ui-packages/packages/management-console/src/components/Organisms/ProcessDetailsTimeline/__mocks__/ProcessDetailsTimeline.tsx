import React from 'react';

const MockedProcessDetailsTimeline = (): React.ReactElement => {
  return <></>;
};

export default MockedProcessDetailsTimeline;
