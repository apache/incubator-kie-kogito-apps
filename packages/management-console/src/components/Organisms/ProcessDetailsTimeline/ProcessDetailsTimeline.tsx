import { TimeAgo } from '@n1ru4l/react-time-ago';
import { Card, CardBody, CardFooter, CardHeader } from '@patternfly/react-core';
import { ServicesIcon, UserIcon } from '@patternfly/react-icons'
import React from 'react';
import './ProcessDetailsTimeline.css';

const ProcessDetailsTimeline = ({ loading, data }) => {

  return (
    <Card>
      <CardHeader>Timeline</CardHeader>
      <CardBody>
        <div className="timeline-container">
          {!loading ? (
            data[0].nodes.map((content, index) => {
              return (
                <div className="timeline-item" key={index}>
                  <div className="timeline-item-content">
                    <TimeAgo date={new Date(`${content.exit}`)} render={({ error, value }) => <span>{value}</span>} />
                    <p>{content.name}</p>
                    <span className="circle">
                      {content.type === 'HumanTaskNode' ? (
                        <UserIcon className="processdetailstimetine-iconstyle" />
                      ) : (
                        <ServicesIcon className="processdetailstimetine-iconstyle" />
                      )}{' '}
                    </span>
                  </div>
                </div>
              );
            })
          ) : (
            <p>loading...</p>
          )}
        </div>
      </CardBody>
    </Card>
  );
};

export default ProcessDetailsTimeline;
