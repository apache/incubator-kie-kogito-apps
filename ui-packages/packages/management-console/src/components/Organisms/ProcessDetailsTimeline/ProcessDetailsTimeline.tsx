import Moment from 'react-moment';
import {
  Card,
  CardBody,
  CardHeader,
  Title,
  Text,
  TextContent,
  TextVariants,
  Split,
  SplitItem,
  Stack,
  Dropdown,
  KebabToggle,
  DropdownItem,
  Tooltip
} from '@patternfly/react-core';
import {
  UserIcon,
  CheckCircleIcon,
  ErrorCircleOIcon,
  OnRunningIcon
} from '@patternfly/react-icons';
import React, { useState } from 'react';
import './ProcessDetailsTimeline.css';
import { GraphQL } from '@kogito-apps/common';
import {
  handleRetry,
  handleSkip,
  handleNodeInstanceRetrigger,
  setTitle,
  handleNodeInstanceCancel
} from '../../../utils/Utils';
import ProcessInstance = GraphQL.ProcessInstance;
import ProcessListModal from '../../Atoms/ProcessListModal/ProcessListModal';

export interface IOwnProps {
  data: Pick<
    ProcessInstance,
    'id' | 'nodes' | 'addons' | 'error' | 'serviceUrl' | 'processId' | 'state'
  >;
}
enum TitleType {
  SUCCESS = 'success',
  FAILURE = 'failure'
}
const ProcessDetailsTimeline: React.FC<IOwnProps> = ({ data }) => {
  const [kebabOpenArray, setKebabOpenArray] = useState([]);
  const [modalTitle, setModalTitle] = useState<string>('');
  const [titleType, setTitleType] = useState<string>('');
  const [modalContent, setModalContent] = useState<string>('');
  const [isModalOpen, setIsModalOpen] = useState<boolean>(false);
  const ignoredNodeTypes = ['Join', 'Split', 'EndNode'];

  const onKebabToggle = (isOpen: boolean, id) => {
    if (isOpen) {
      setKebabOpenArray([...kebabOpenArray, id]);
    } else {
      onDropdownSelect(id);
    }
  };

  const onDropdownSelect = id => {
    const tempKebabArray = [...kebabOpenArray];
    const index = tempKebabArray.indexOf(id);
    tempKebabArray.splice(index, 1);
    setKebabOpenArray(tempKebabArray);
  };

  const handleModalToggle = () => {
    setIsModalOpen(!isModalOpen);
  };

  const onShowMessage = (
    title: string,
    content: string,
    type: TitleType
  ): void => {
    setTitleType(type);
    setModalTitle(title);
    setModalContent(content);
    handleModalToggle();
  };
  const dropdownItems = (processInstanceData, node) => {
    if (
      processInstanceData.error &&
      node.definitionId === processInstanceData.error.nodeDefinitionId
    ) {
      return [
        <DropdownItem
          key="retry"
          component="button"
          onClick={() =>
            handleRetry(
              processInstanceData,
              () =>
                onShowMessage(
                  'Retry operation',
                  `The node ${node.name} was successfully re-executed.`,
                  TitleType.SUCCESS
                ),
              (errorMessage: string) =>
                onShowMessage(
                  'Retry operation',
                  `The node ${node.name} failed to re-execute. Message: ${errorMessage}`,
                  TitleType.FAILURE
                )
            )
          }
        >
          Retry
        </DropdownItem>,
        <DropdownItem
          key="skip"
          component="button"
          onClick={() =>
            handleSkip(
              processInstanceData,
              () =>
                onShowMessage(
                  'Skip operation',
                  `The node ${node.name} was successfully skipped.`,
                  TitleType.SUCCESS
                ),
              (errorMessage: string) =>
                onShowMessage(
                  'Skip operation',
                  `The node ${node.name} failed to skip. Message: ${errorMessage}`,
                  TitleType.FAILURE
                )
            )
          }
        >
          Skip
        </DropdownItem>
      ];
    } else if (node.exit === null && !ignoredNodeTypes.includes(node.type)) {
      return [
        <DropdownItem
          key="retrigger"
          component="button"
          onClick={() =>
            handleNodeInstanceRetrigger(
              processInstanceData,
              node,
              () =>
                onShowMessage(
                  'Node retrigger operation',
                  `The node ${node.name} was successfully retriggered.`,
                  TitleType.SUCCESS
                ),
              (errorMessage: string) =>
                onShowMessage(
                  'Node retrigger operation',
                  `The node ${node.name} failed to retrigger. Message: ${errorMessage}`,
                  TitleType.FAILURE
                )
            )
          }
        >
          Retrigger node
        </DropdownItem>,
        <DropdownItem
          key="cancel"
          component="button"
          onClick={() =>
            handleNodeInstanceCancel(
              processInstanceData,
              node,
              () =>
                onShowMessage(
                  'Node cancel operation',
                  `The node ${node.name} was successfully canceled.`,
                  TitleType.SUCCESS
                ),
              (errorMessage: string) =>
                onShowMessage(
                  'Node cancel operation',
                  `The node ${node.name} failed to cancel. Message: ${errorMessage}`,
                  TitleType.FAILURE
                )
            )
          }
        >
          Cancel node
        </DropdownItem>
      ];
    } else {
      return [];
    }
  };
  const processManagementKebabButtons = (node, index) => {
    const dropdownItemsValue = dropdownItems(data, node);
    if (
      data.addons.includes('process-management') &&
      data.serviceUrl !== null &&
      dropdownItemsValue.length !== 0
    ) {
      return (
        <Dropdown
          onSelect={() => onDropdownSelect('timeline-kebab-toggle-' + index)}
          toggle={
            <KebabToggle
              onToggle={isOpen =>
                onKebabToggle(isOpen, 'timeline-kebab-toggle-' + index)
              }
              id={'timeline-kebab-toggle-' + index}
            />
          }
          position="right"
          isOpen={kebabOpenArray.includes('timeline-kebab-toggle-' + index)}
          isPlain
          dropdownItems={dropdownItemsValue}
        />
      );
    }
  };

  return (
    <Card>
      <ProcessListModal
        isModalOpen={isModalOpen}
        handleModalToggle={handleModalToggle}
        checkedArray={data && [data.state]}
        modalTitle={setTitle(titleType, modalTitle)}
        modalContent={modalContent}
      />
      <CardHeader>
        <Title headingLevel="h3" size="xl">
          Timeline
        </Title>
      </CardHeader>
      <CardBody>
        <Stack gutter="md" className="kogito-management-console--timeline">
          {data.nodes &&
            data.nodes.map((content, idx) => {
              return (
                <Split
                  gutter={'sm'}
                  className={'kogito-management-console--timeline-item'}
                  key={content.id}
                >
                  <SplitItem>
                    {
                      <>
                        {data.error &&
                        content.definitionId === data.error.nodeDefinitionId ? (
                          <Tooltip content={data.error.message}>
                            <ErrorCircleOIcon
                              color="var(--pf-global--danger-color--100)"
                              className="kogito-management-console--timeline-status"
                            />
                          </Tooltip>
                        ) : content.exit === null ? (
                          <Tooltip content={'Active'}>
                            <OnRunningIcon className="kogito-management-console--timeline-status" />
                          </Tooltip>
                        ) : (
                          <Tooltip content={'Completed'}>
                            <CheckCircleIcon
                              color="var(--pf-global--success-color--100)"
                              className="kogito-management-console--timeline-status"
                            />
                          </Tooltip>
                        )}
                      </>
                    }
                  </SplitItem>
                  <SplitItem isFilled>
                    <TextContent>
                      <Text component={TextVariants.p}>
                        {content.name}
                        <span>
                          {content.type === 'HumanTaskNode' && (
                            <Tooltip content={'Human task'}>
                              <UserIcon
                                className="pf-u-ml-sm"
                                color="var(--pf-global--icon--Color--light)"
                              />
                            </Tooltip>
                          )}
                        </span>
                        <Text component={TextVariants.small}>
                          {content.exit === null ? (
                            'Active'
                          ) : (
                            <Moment fromNow>
                              {new Date(`${content.exit}`)}
                            </Moment>
                          )}
                        </Text>
                      </Text>
                    </TextContent>
                  </SplitItem>
                  <SplitItem>
                    {processManagementKebabButtons(content, idx)}
                  </SplitItem>
                </Split>
              );
            })}
        </Stack>
      </CardBody>
    </Card>
  );
};

export default ProcessDetailsTimeline;
