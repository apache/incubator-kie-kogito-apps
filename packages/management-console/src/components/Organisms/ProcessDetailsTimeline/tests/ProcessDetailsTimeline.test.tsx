import React from 'react';
import { shallow } from 'enzyme';
import ProcessDetailsTimeline from '../ProcessDetailsTimeline';

const props = {
    loading: false,
    data: [{
        nodes: [{
            name: "Book flight",
            definitionId: "ServiceTask_1",
            id: "2f588da5-a323-4111-9017-3093ef9319d1",
            enter: "2019-10-22T04:43:01.144Z",
            exit: "2019-10-22T04:43:01.144Z",
            type: "WorkItemNode"
        }]
    }]
}


const props2 = {
    loading: true,
    data: [{}]
}

const props3 = {
    loading: false,
    data: [{
        nodes: [{
            name: "Book flight",
            definitionId: "ServiceTask_1",
            id: "2f588da5-a323-4111-9017-3093ef9319d1",
            enter: "2019-10-22T04:43:01.144Z",
            exit: "2019-10-22T04:43:01.144Z",
            type: "HumanTaskNode"
        }]
    }]
}
describe('Process Details Timeline component', () => {
    it('Sample test case', () => {
        const wrapper = shallow(<ProcessDetailsTimeline {...props}/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.circle').length).toEqual(1);
      });
    it('Sample test case', () => {
        const wrapper = shallow(<ProcessDetailsTimeline {...props2}/>);
        expect(wrapper.find('.circle').length).toEqual(0);
        
    });
    it('test human task', () => {
        const wrapper = shallow(<ProcessDetailsTimeline {...props3}/>);
        expect(wrapper).toMatchSnapshot();
        
    });
})

