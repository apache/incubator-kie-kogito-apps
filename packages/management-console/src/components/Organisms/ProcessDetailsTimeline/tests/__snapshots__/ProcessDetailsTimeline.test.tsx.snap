// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Process Details Timeline component Sample test case 1`] = `
<Card>
  <CardHeader>
    <Title
      headingLevel="h3"
      size="xl"
    >
      Timeline
    </Title>
  </CardHeader>
  <CardBody>
    <div
      className="timeline-container"
    >
      <div
        className="timeline-item"
        key="2f588da5-a323-4111-9017-3093ef9319d1"
      >
        <div
          className="timeline-item-content"
        >
          <TimeAgo
            date={2019-10-22T04:43:01.144Z}
            formatter={[Function]}
            render={[Function]}
          />
          <p>
            Book flight
          </p>
          <span
            className="circle"
          >
            <ServicesIcon
              className="processdetailstimetine-iconstyle"
              color="currentColor"
              noVerticalAlign={false}
              size="sm"
              title={null}
            />
             
          </span>
        </div>
      </div>
    </div>
  </CardBody>
</Card>
`;

exports[`Process Details Timeline component test human task 1`] = `
<Card>
  <CardHeader>
    <Title
      headingLevel="h3"
      size="xl"
    >
      Timeline
    </Title>
  </CardHeader>
  <CardBody>
    <div
      className="timeline-container"
    >
      <div
        className="timeline-item"
        key="2f588da5-a323-4111-9017-3093ef9319d1"
      >
        <div
          className="timeline-item-content"
        >
          <TimeAgo
            date={2019-10-22T04:43:01.144Z}
            formatter={[Function]}
            render={[Function]}
          />
          <p>
            Book flight
          </p>
          <span
            className="circle"
          >
            <UserIcon
              className="processdetailstimetine-iconstyle"
              color="currentColor"
              noVerticalAlign={false}
              size="sm"
              title={null}
            />
             
          </span>
        </div>
      </div>
    </div>
  </CardBody>
</Card>
`;
